### DISCUSSION TITLE
chapter_searching/replace_linear_by_hashing/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#33] [Announcements]
# chapter_searching/replace_linear_by_hashing/

一本动画图解、能运行、可提问的数据结构与算法入门书

https://www.hello-algo.com/chapter_searching/replace_linear_by_hashing/

**Comment by hatkd** [comment]
很谢谢作者

**Reply by krahets** [reply]
感谢鼓励！

**Comment by Wonderdch** [comment]
建议将“两数之和”的题目也放进来，免得跳转过去看分心

**Reply by JLA97** [reply]
本书的定位应该不是那种带着刷算法的，为了保持简洁，放个链接应该够了

**Reply by FangYuan33** [reply]
我觉得应该放进来，至少我读到这儿的时候在下意识的找题干，如果是点击链接还是觉得有些麻烦

**Reply by krahets** [reply]
🫡收到已加～

**Comment by paulxu001** [comment]
dic.containsKey函数的最差时间复杂度不是O(1)

**Reply by krahets** [reply]
是的，在某些条件下会劣化～但在做题时，我们一般会认为哈希表时间复杂度 O(1) 。

**Comment by lirantongxue** [comment]
元旦过后,遇到最好的礼物。我真心觉得这么好的入门工具书,首先应该让人读的尽兴。而不是半遮半掩,我希望后面那些有规划没有写出来的内容抓紧和我们读者见面。否则很扫兴呀。

**Reply by krahets** [reply]
谢谢鼓励！在加紧更新，请再给我一些时间。PS：没有半遮半掩，感觉很努力地在写了哈哈

**Reply by xlcodes** [reply]
作者大大加油，算法小白一枚持续关注ing...

**Comment by Chilled-Coke** [comment]
为啥程序在C语言那一栏显示不了？

**Reply by krahets** [reply]
C 语言代码暂时还在测试，等测试好会加入～！

**Reply by Miyolovecode** [reply]
期待期待！！

**Comment by zh-h** [comment]
两数之和那道题，有点背单词 abandon 的感觉。

**Reply by krahets** [reply]
有那味了

**Comment by dengyunsheng250** [comment]
没有C的代码吗？

**Reply by krahets** [reply]
抱歉！ C 的代码还在翻译+测试中，可能暂时只能用 C++ 来代替

**Reply by Chilled-Coke** [reply]
作者大大加油！

**Comment by happykala** [comment]
作者你好，示例代码对于处理nums列表中存在多个满足两个数值之和为target的情况是不是解决不了呢？？

**Reply by krahets** [reply]
是的，这道题要求返回第一对即可

**Comment by TDB001** [comment]
```cpp
// C++
vector<int> vec {1, 2, 3, 4};
cout << twoSumBruteForce(vec, 5) << endl;  // [0, 3]
cout << twoSumHashTable(vec, 5) << endl;   // [1, 2]
```

```swift
// Swift
let nums = [1, 2, 3, 4]
print(twoSumBruteForce(nums: nums, target: 5)) // [0, 3]
print(twoSumHashTable(nums: nums, target: 5))  // [1, 2]
```

k神，两个方法结果不同，twoSumHashTable 这个取到的不是第1个下标。

**Reply by krahets** [reply]
这个没关系，题目只要求输出任一满足条件的组合即可。

由于两个方法的遍历方式不同，因此它们找到解的顺序是不同的。暴力法是普通的双层遍历，总会优先找到包含靠左边元素的组合。哈希表法是一次遍历，在你展示的示例中就会先找到 `[1, 2]` 这个组合。

**Reply by TDB001** [reply]
👌

**Comment by TangJun815** [comment]
测试了一下，还是需要具体问题具体分析，如下情况线性查找就比哈希好。


~~~code csharp
C#测试代码
//准备
var sw = new Stopwatch();
var nums = Enumerable.Range(1,int.MaxValue/2).ToArray();

//暴力枚举
sw.Start();
var s = twoSumBruteForce(nums, 999999999);
sw.Stop();
Console.WriteLine($"暴力枚举用时:{sw.Elapsed.TotalSeconds}秒");
foreach (var i in s)
{
	Console.WriteLine($"索引：{i}，值：{nums[i]}");
}

//辅助哈希
sw.Start();
var s2 = twoSumHashTable(nums, 999999999);
sw.Stop();
Console.WriteLine($"辅助哈希表用时:{sw.Elapsed.TotalSeconds}秒");
foreach (var i in s2)
{
	Console.WriteLine($"索引：{i}，值：{nums[i]}");
}
~~~
输出结果：
暴力枚举用时:1.7970771秒
索引：0，值：1
索引：999999997，值：999999998
辅助哈希表用时:51.2555588秒
索引：499999998，值：499999999
索引：499999999，值：500000000

**Comment by fiona-liu12** [comment]
如果能给出对应力扣上可以训练同类型的题目就好了

**Reply by krahets** [reply]
推荐这个[题单](https://leetcode.cn/studyplan/selected-coding-interview/)

**Reply by Cupcc** [reply]
希望能添加相应题目的链接

**Reply by WheatJack** [reply]
https://leetcode.cn/problems/two-sum/

**Comment by newcakes** [comment]
@krahets佬，为什么C++哈希查找的算法时间复杂度是O(n)呀！在使用find()这个函数的时候不是需要遍历哈希表吗？时间复杂度是O(n),n次不就是O(n2)吗？
还是说哈希的find()函数的时间复杂度是O(1)。

**Reply by krahets** [reply]
`unordered_map` 是哈希表，它的 `find()` 的时间复杂度是 O(1) 哈

**Reply by newcakes** [reply]
好的，谢谢

**Comment by ChinaYong** [comment]
问一下，为什么要用诸如
```java
Map<Integer, Integer> dic = new HashMap<>();
```
的语句啊？相比
```java
HashMap<Integer, Integer> dic = new HashMap<>();
```
有什么优点吗？

**Reply by krahets** [reply]
好问题！使用接口类型作为引用类型可以使代码的意图更加清晰，它表明我们只打算使用接口提供的方法，而不是特定实现的额外方法。  从逻辑上讲，我们想用的是一个 Map 来实现映射功能，而出于效率考虑，使用 HashMap 作为这个 Map 的实现。

**Reply by HangSeven** [reply]
这是不是就是多态，编译看左边，运行看右边

**Reply by HuaWang135608** [reply]
因为 Map 是接口，HashMap 是 Map 接口的实现类；这个就是面向接口编程的习惯，感兴趣可以看一下设计模式中的开闭原则和依赖倒置原则。

**Comment by Terran0629** [comment]
leetcode第一题哈哈哈哈大家的入门题

**Reply by qw124341789** [reply]
要是力扣太难去刷PTA吧

**Comment by codeberg-user** [comment]
为什么暴力时不先把target - nums[i]提出来? *ns级优化*

**Reply by Lirous587** [reply]
怎么个优化法？提出来还不是要再来一个循环吗

**Comment by RENHJa** [comment]
请问，c语言暴力枚举中函数第四个参数`int *returnSize`是什么作用？

**Reply by krahets** [reply]
用于记录 `res` 的长度，函数返回后需要用到

**Reply by RENHJa** [reply]
那为什么记录`res`的长度是 int* 类型而不是 int 型呢？

**Reply by codeberg-user** [reply]
可能是int *改变实参,int 只能改变形参

**Reply by Lirous587** [reply]
因为返回数组之后你需要一个数组的lens 数组不知道lens的话这个数组也就没意义了  而c里面计算lens是很不方便的 所以直接传一个int* returnSize就很方便了 或者你可以传一个结构体的指针 里面带有
```c
typedef struct {
    int returnSize;
    int * res ;
} ReturnData;
```
反正数组必须要有它的lens 至于方法无所谓的

**Comment by Tiro-Wang** [comment]
c#中这段代码
```
        if (dic.ContainsKey(target - nums[i])) {
            return [dic[target - nums[i]], i];
        }
        dic.Add(nums[i], i);
```
改变一下顺序也是可以的，
```
        dic.Add(nums[i], i);
        if (dic.ContainsKey(target - nums[i])) {
            return [dic[target - nums[i]], i];
        }
```
目前想不懂源代码的写法是不是更有优势呢。

**Reply by KLYkl** [reply]
如果你改了，刚刚好遇到一个索引的数是target的两倍，但如果是题目要求不能是数组的同一个索引，请问阁下如何应对？哈哈哈

**Reply by KLYkl** [reply]
而根据作者给出的解法，默认是不能是同一个数组索引的元素。

**Comment by Re-xiguan** [comment]
```cpp
/* 方法二：辅助哈希表 */
vector<int> twoSumHashTable(vector<int> &nums, int target) {
    int size = nums.size();
    // 辅助哈希表，空间复杂度为 O(n)
    unordered_map<int, int> dic;
    // 单层循环，时间复杂度为 O(n)
    for (int i = 0; i < size; i++) {
        if (dic.find(target - nums[i]) != dic.end()) {
            return {dic[target - nums[i]], i};
        }
        dic.emplace(nums[i], i);
    }
    return {};
}
```
辅助哈希表中的查找和维护时间复杂度不是log（n）吗

**Reply by Re-xiguan** [reply]
借助哈希表搜索时间复杂度为什么会降到n啊

**Reply by wqc18737346950** [reply]
我的理解是，hash表中增删改操作时间复杂度都是O(1)，然后只有一层循环，所以是O(n)

**Reply by 1435646097** [reply]
是的，hash表中增删改操作时间复杂度都是O(1)，只用管外层循环就可以了

**Comment by sunshinesDL** [comment]
K大，举一反三，假定题中的两个元素可以重复，例如：
```
    nums = [2, 7, 11, 15]
    target = 14
```
可以返回 [1, 1]作为结果，两种解法该如何更改呢？

**Reply by krahets** [reply]
Hi，由于是两个元素，因此遍历每个元素 n 试一下 n*n 是否可以就行了，仍然是一个 O(n) 的操作

**Reply by KLYkl** [reply]
将dic插入的操作位置放在if判断之前即可。

**Comment by niyongsheng** [comment]
补充一个排序+双指针策略，空间复杂度O(1) 时间复杂度O(nlogn)
```java
int[] findTwoSum(int[] nums, int target) {
    Arrays.sort(nums); // 对数组排序
    int left = 0;
    int right = nums.length - 1;
    while (left < right) {
        int sum = nums[left] + nums[right];
        if (sum == target) {
            return new int[]{left, right};
        } else if (sum < target) {
            left++;
        } else {
            right--;
        }
    }
    return null;
```

**Reply by xbuffer** [reply]
你的算法前提是数组要有序

**Reply by xbuffer** [reply]
数组排序的时间复杂度也要加上才行

**Reply by niyongsheng** [reply]
> 根据大 O 符号的定义，它描述算法运行时间的上限。在多个复杂度项相加的情况下，通常只保留最高阶项。

`O(n log n)(排序) + O(n)(双指针)`由于排序是主要的复杂度，因此总的时间复杂度是` O(n log n)`

#### 不过排序后拿到索引已经不是原数组的索引了，可以创建一个辅助二维数组来存储元素和原始索引，在创建的同时使用归并排序。
* 时间复杂度：O(n log n)
* 空间复杂度：O(log n)

还是哈希表策略更优，只是提供一个思路；

**Comment by lt421** [comment]
001 两数之和！

**Comment by ghost** [comment]
two_sum.c中针对uthash.h提供的函数的使用是否有错误？查阅文档，似乎HASH_ADD_INT（）等接收的第一个参数均为指向HashTable指针的指针

**Comment by LifeIsTerrible** [comment]
nice

**Comment by lexign** [comment]
发现一个小bug，图10-9的“7”行加法算错了，“7+15”=22不是23😄

**Reply by krahets** [reply]
谢谢指正！

**Comment by moeselou** [comment]
感觉这个子章节出现的有点突然，脑回路跳跃有点大，仅用这语”我们借助一个算法题来加深理解。“就转向了，一开始给我看的云里雾里，建议更明确的说明下才更好，不然会以为承接二分查找的哈希优化呢

**Comment by elk-9527** [comment]
力扣第一题“两数之和”

**Comment by xiaozhuge123** [comment]
day08

**Comment by mygit10636** [comment]
这是怎么想出来用哈希表的，有点八竿子打不着的感觉

**Comment by Ivy129** [comment]
不太明白为什么最后一个哈希查找，可以找到目标值

**Reply by ashjiege** [reply]
其实它就是做一个将数组遍历存入哈希表的过程，只不过依靠这个过程来实现两数之和，简单点说就是3+7=10和7+3=10是一样的，在遍历的过程中先存入3还是先存入7都不影响，只需要后面遍历到另一个值的时候，就能找到先存入的那个值

**Comment by jyqs9** [comment]
看了半天才看懂，我成`rz`了

**Comment by Randomlyu** [comment]
这个问题不就是尺取法嘛，用双指针就解决了

**Comment by Wei-Jie-YUAN** [comment]
LeetCode第一题，两数之和！
